homeassistant/components/yeelight/__init__.py

```diff
@@ -262,7 +271,27 @@ def update(self):
             self._available = True
         except BulbException as ex:
             if self._available:  # just inform once
-                _LOGGER.error("Unable to update bulb status: %s", ex)
+                _LOGGER.error("Unable to update device %s, %s: %s",
+                              self.ipaddr, self.name, ex)
             self._available = False
 
+        return self._available
+
+    def _initialize_if_not_already(self):
+        if self._initialized:
+            return
+
+        self._initialized = True
+        dispatcher_send(self._hass, DEVICE_INITIALIZED, self.ipaddr)
+
+    def update(self):
+        """Update device properties and send data updated signal."""
+        if self._update_properties():
+            self._initialize_if_not_already()
+
         dispatcher_send(self._hass, DATA_UPDATED.format(self._ipaddr))
+
+    def setup(self):
+        """Initialize device and send initialized signal."""
+        if self._update_properties() or self.model:
```

Comment:
Should the one above also check for `self.model` like this?

Reply:
I'm not sure which one ? Can you link to code line ?

Reply:
Line 289, in `update()`. I think this code structure is a bit confusing and could be made easier to follow by restructuring it somehow.

Reply:
Model was only needed in setup, when update fails, but we have model declared, we can build correct Yeelight class. I've simplified this code, I hope it easier to follow now.